### PR TITLE
Fenics 2019.1.0 focal tf

### PR DESCRIPTION
**Pull request by mspieg**
New docker files for building Terraferma/base, dev-env, dev on a phusion 20.04 base

### CODE REVIEW

File: dockerfiles/dev-env/Dockerfile [hunk 1]
```diff
@@ -143,23 +151,24 @@ ENV PETSC_PREFIX /usr/local/petsc
...
-                        --download-hypre=1 \
+                        --download-hypre=/tmp/v2.15.1.tar.gz \
                         --download-ml=1 \
                         --download-parmetis=1 \
                         --download-metis=1 \
                         --download-blacs=1 \
                         --download-scalapack=1 \
                         --download-mumps=1 \
-                        --download-petsc4py=1 \
-                        --with-debugging=0 \
```
Comment (cianwilson) [inline dockerfiles/dev-env/Dockerfile:162]: How come you removed these?

File: dockerfiles/dev/Dockerfile [hunk 2]
```diff
@@ -26,25 +26,26 @@
...
 MAINTAINER terraferma <terraferma@lists.columbia.edu>
 
 USER root
 WORKDIR /tmp
 
+
 # install terraferma
-ENV TF_PREFIX /usr/local/terraferma
-RUN export PYTHONPATH=$FENICS_PREFIX/lib/python3.6/site-packages && \
+ENV TF_PREFIX=/usr/local/terraferma
```
Comment (cianwilson) [inline dockerfiles/dev/Dockerfile]: Why this change of syntax?

File: dockerfiles/dev/Dockerfile [hunk 3]
```diff
@@ -26,25 +26,26 @@
...
-RUN export PYTHONPATH=$FENICS_PREFIX/lib/python3.6/site-packages && \
+ENV TF_PREFIX=/usr/local/terraferma 
+RUN export PYTHONPATH=$FENICS_PREFIX/lib/python3.8/site-packages && \
     export PETSC_DIR=$PETSC_PREFIX && \
     export PATH=$FENICS_PREFIX/bin:$PATH && \
     export LD_LIBRARY_PATH=$FENICS_PREFIX/lib:$LD_LIBRARY_PATH && \
     export PKG_CONFIG_PATH=$FENICS_PREFIX/lib/pkgconfig:$PKG_CONFIG_PATH && \
     export SPUD_DIR=$SPUD_PREFIX && \
-    git clone -b master-2019.1.0 -q https://github.com/TerraFERMA/TerraFERMA.git && \
+    git clone -b mspieg/master-2019.1.0-update-fixes -q https://github.com/TerraFERMA/TerraFERMA.git && \
```
Comment (cianwilson) [inline dockerfiles/dev/Dockerfile]: I think after yesterday's PR merge into TF this can be reverted?

File: dockerfiles/dev/Dockerfile [hunk 4]
```diff
@@ -26,25 +26,26 @@
...
     cd TerraFERMA && \
     mkdir build && \
     cd build && \
-    cmake -DCMAKE_INSTALL_PREFIX=${TF_PREFIX} \
+    cmake -DCMAKE_INSTALL_PREFIX=${TF_PREFIX} -DCMAKE_BUILD_TYPE=Release  \
```
Comment (cianwilson) [inline dockerfiles/dev/Dockerfile]: Pedantic but could this be split up over multiple lines for consistency with the other cmake commands?

File: dockerfiles/dev-env/Dockerfile [hunk 5]
```diff
@@ -30,7 +30,7 @@
 #
 
 # build from the base image
-FROM terraferma/base:fenics-2019.1.0
+FROM terraferma/base:fenics-2019.1.0-focal
```
Comment (cianwilson) [inline dockerfiles/dev-env/Dockerfile:33]: Think this can just be renamed to fenics-2019.1.0 and take over from the current version.

File: dockerfiles/dev/Dockerfile [hunk 6]
```diff
@@ -26,25 +26,26 @@
 #
 
 # build using the TerraFERMA development environment
-FROM terraferma/dev-env:fenics-2019.1.0
+FROM terraferma/dev-env:fenics-2019.1.0-focal
```
Comment (cianwilson) [inline dockerfiles/dev/Dockerfile:29]: Can similarly be renamed to just fenics-2019.1.0

**Review by cianwilson — CHANGES_REQUESTED** [review]
Looks great, thanks for updating.

Some minor fixes in comments.

File: dockerfiles/dev-env/Dockerfile (hunk 1 above)
Comment (mspieg) [reply dockerfiles/dev-env/Dockerfile:162]: --with-debugging just got moved up in the list (or should have)

and petsc4py wouldn't build under Python3.8 (and didn't seem to be needed for the rest of the stack) so I just got lazy and took it out of the petsc build

File: dockerfiles/dev/Dockerfile (hunk 2 above)
Comment (mspieg) [reply dockerfiles/dev/Dockerfile]: Not sure what's going on here,  I think you just need to look at the whole docker file.  I needed to change the python path to accomodate python3.8.  but that line should be in there below but read 

RUN export PYTHONPATH=$FENICS_PREFIX/lib/python3.8/site-packages && \

File: dockerfiles/dev/Dockerfile (hunk 3 above)
Comment (mspieg) [reply dockerfiles/dev/Dockerfile]: probably...
I forgot about this,  yes my update-fixes were just the 2to3 fixes for updateshml, updatetfml so if they're in their definitely revert.

File: dockerfiles/dev/Dockerfile (hunk 4 above)
Comment (mspieg) [reply dockerfiles/dev/Dockerfile]: as you like...but the Release really needs to be there

File: dockerfiles/dev-env/Dockerfile (hunk 5 above)
Comment (mspieg) [reply dockerfiles/dev-env/Dockerfile:33]: You're call entirely.  I just  didn't want to suprise someone who had an old version and suddenly we swapped out a whole OS version (plus two releases of python3 on them)

File: dockerfiles/dev/Dockerfile (hunk 6 above)
Comment (mspieg) [reply dockerfiles/dev/Dockerfile:29]: ditto

File: dockerfiles/dev-env/Dockerfile (hunk 5 above)
Comment (cianwilson) [reply dockerfiles/dev-env/Dockerfile:33]: Good point.  Agreed, let's leave the `-focal` tag.

File: dockerfiles/dev-env/Dockerfile (hunk 1 above)
Comment (cianwilson) [reply dockerfiles/dev-env/Dockerfile:162]: OK, since this is going to live as a separate tag it's not important.  Petsc4py is sometimes useful to have but it's fine if it's removed for now.

Apologies hadn't spotted the debugging move.

File: dockerfiles/dev/Dockerfile (hunk 6 above)
Comment (cianwilson) [reply dockerfiles/dev/Dockerfile:29]: ditto ditto

**Review by cianwilson — APPROVED** [review]
Looks good, thanks for the changes!
